src/baseline/cannon/codegen.rs

```diff
@@ -0,0 +1,139 @@
+use crate::class::TypeParams;
+use crate::ctxt::VM;
+use crate::ctxt::{Fct, FctSrc};
+use crate::masm::*;
+
+use crate::bytecode::astgen::generate_fct;
+use crate::bytecode::generate::{BytecodeFunction, Register};
+use crate::bytecode::opcode::Bytecode;
+
+pub struct CannonCodeGen<'a, 'ast: 'a> {
+    pub vm: &'a VM<'ast>,
+    pub fct: &'a Fct<'ast>,
+    pub ast: &'ast Function,
+    pub asm: BaselineAssembler<'a, 'ast>,
+    pub scopes: Scopes,
+    pub src: &'a mut FctSrc,
+
+    pub lbl_break: Option<Label>,
+    pub lbl_continue: Option<Label>,
+
+    // stores all active finally blocks
+    pub active_finallys: Vec<&'ast Stmt>,
+
+    // label to jump instead of emitting epilog for return
+    // needed for return's in finally blocks
+    // return in finally needs to execute to next finally block and not
+    // leave the current function
+    pub lbl_return: Option<Label>,
+
+    // length of active_finallys in last loop
+    // default: 0
+    // break/continue need to emit finally blocks up to the last loop
+    // see tests/finally/break-while.dora
+    pub active_loop: Option<usize>,
+
+    // upper length of active_finallys in emitting finally-blocks for break/continue
+    // default: active_finallys.len()
+    // break/continue needs to execute finally-blocks in loop, return in these blocks
+    // would dump all active_finally-entries from the loop but we need an upper bound.
+    // see emit_finallys_within_loop and tests/finally/continue-return.dora
+    pub active_upper: Option<usize>,
+
+    pub cls_type_params: &'a TypeParams,
+    pub fct_type_params: &'a TypeParams,
+}
+
+impl<'a, 'ast> CannonCodeGen<'a, 'ast>
+where
+    'ast: 'a,
+{
+    fn emit_prolog(&mut self, bytecode: &BytecodeFunction) {
+        self.asm.prolog(bytecode.stacksize());
+        self.asm.emit_comment(Comment::Lit("prolog end"));
+        self.asm.emit_comment(Comment::Newline);
+    }
+
+    fn emit_epilog(&mut self, bytecode: &BytecodeFunction) {
+        self.asm.emit_comment(Comment::Newline);
+        self.asm.emit_comment(Comment::Lit("epilog"));
+
+        let polling_page = self.vm.polling_page.addr();
+        self.asm
+            .epilog_with_polling(bytecode.stacksize(), polling_page);
+    }
+
+    fn emit_const_bool(&mut self, bytecode: &BytecodeFunction, dest: &Register, bool_const: bool) {
+        let Register(index) = *dest;
+        let bytecode_type = bytecode.registers().get(index).expect("register not found");
+        let offset = bytecode.offset().get(index).expect("offset not found");
+
+        if bool_const {
+            self.asm.load_true(REG_RESULT);
+        } else {
+            self.asm.load_false(REG_RESULT);
+        }
+        self.asm
+            .store_mem(bytecode_type.mode(), Mem::Local(*offset), REG_RESULT.into());
+    }
+}
+
+impl<'a, 'ast> CodeGen<'ast> for CannonCodeGen<'a, 'ast> {
+    fn generate(mut self) -> JitBaselineFct {
+        let bytecode = generate_fct(
```

Comment:
I think it makes sense to move the bytecode generation out of CannonCodeGen. Its responsibility is to generate Machinecode from the bytecode, but not to generate the bytecode. Then you could add the bytecode as a field in CannonCodeGen and you don't need to pass it to each function as argument.

Reply:
I agree. What do you think about a factory or something similar for the code generators? Then we could move the creation of the baseline compiler away from `baseline/codegen.rs` and reduce the boiler plate code. 
I would do this in another PR, so we could merge this on. Obviously, only if this is OK for you.